File: shifai-ios/ShifAI/Data/Encryption/EncryptionManager.swift
import Foundation
import CryptoKit
import CommonCrypto

// MARK: - Encryption Manager
// AES-256-GCM encryption for sync data, PBKDF2 key derivation
// Spike S0-1: IMPLEMENTED — zero-knowledge architecture

/// Manages all encryption operations following zero-knowledge architecture
final class EncryptionManager {

    static let shared = EncryptionManager()

    // MARK: - Errors

    enum EncryptionError: Error, LocalizedError {
        case keyDerivationFailed
        case encryptionFailed
        case decryptionFailed
        case invalidData
        case keychainAccessFailed
        case saltGenerationFailed
        case invalidKeyLength

        var errorDescription: String? {
            switch self {
            case .keyDerivationFailed: return "Échec de la dérivation de clé"
            case .encryptionFailed: return "Échec du chiffrement"
            case .decryptionFailed: return "Échec du déchiffrement"
            case .invalidData: return "Données invalides"
            case .keychainAccessFailed: return "Accès Keychain échoué"
            case .saltGenerationFailed: return "Génération du sel échouée"
            case .invalidKeyLength: return "Longueur de clé invalide"
            }
        }
    }

    // MARK: - Key Management

    /// Derive master key from user PIN/biometric-bound secret using PBKDF2
    /// - Parameters:
    ///   - password: User's PIN or biometric-derived secret
    ///   - salt: Random salt (stored in Keychain)
    /// - Returns: Derived key (256-bit / 32 bytes)
    func deriveMasterKey(from password: Data, salt: Data) throws -> Data {
        let iterations = AppConfig.Encryption.pbkdf2Iterations
        let keyLength = AppConfig.Encryption.keyLength / 8 // 256 bits → 32 bytes

        var derivedKey = Data(count: keyLength)

        let result = derivedKey.withUnsafeMutableBytes { derivedKeyBytes in
            password.withUnsafeBytes { passwordBytes in
                salt.withUnsafeBytes { saltBytes in
                    CCKeyDerivationPBKDF(
                        CCPBKDFAlgorithm(kCCPBKDF2),
                        passwordBytes.baseAddress?.assumingMemoryBound(to: Int8.self),
                        password.count,
                        saltBytes.baseAddress?.assumingMemoryBound(to: UInt8.self),
                        salt.count,
                        CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                        UInt32(iterations),
                        derivedKeyBytes.baseAddress?.assumingMemoryBound(to: UInt8.self),
                        keyLength
                    )
                }
            }
        }

        guard result == kCCSuccess else {
            throw EncryptionError.keyDerivationFailed
        }

        return derivedKey
    }

    /// Generate cryptographically secure random salt
    func generateSalt() throws -> Data {
        var salt = Data(count: AppConfig.Encryption.saltLength)
        let result = salt.withUnsafeMutableBytes { buffer in
            SecRandomCopyBytes(kSecRandomDefault, AppConfig.Encryption.saltLength, buffer.baseAddress!)
        }
        guard result == errSecSuccess else {
            throw EncryptionError.saltGenerationFailed
        }
        return salt
    }

    /// Generate a random 256-bit key
    func generateRandomKey() throws -> Data {
        var key = Data(count: 32) // 256 bits
        let result = key.withUnsafeMutableBytes { buffer in
            SecRandomCopyBytes(kSecRandomDefault, 32, buffer.baseAddress!)
        }
        guard result == errSecSuccess else {
            throw EncryptionError.saltGenerationFailed
        }
        return key
    }

    // MARK: - Encryption / Decryption

    /// Encrypt data using AES-256-GCM
    /// - Parameters:
    ///   - data: Plaintext data to encrypt
    ///   - key: 256-bit encryption key (32 bytes)
    /// - Returns: Encrypted data: nonce (12) + ciphertext + tag (16)
    func encrypt(_ data: Data, with key: Data) throws -> Data {
        guard key.count == 32 else {
            throw EncryptionError.invalidKeyLength
        }

        let symmetricKey = SymmetricKey(data: key)

        guard let sealedBox = try? AES.GCM.seal(data, using: symmetricKey) else {
            throw EncryptionError.encryptionFailed
        }

        // Combined: nonce (12 bytes) + ciphertext + tag (16 bytes)
        guard let combined = sealedBox.combined else {
            throw EncryptionError.encryptionFailed
        }

        return combined
    }

    /// Decrypt data using AES-256-GCM
    /// - Parameters:
    ///   - encryptedData: Combined nonce (12) + ciphertext + tag (16)
    ///   - key: 256-bit encryption key (32 bytes)
    /// - Returns: Decrypted plaintext data
    func decrypt(_ encryptedData: Data, with key: Data) throws -> Data {
        guard key.count == 32 else {
            throw EncryptionError.invalidKeyLength
        }

        guard encryptedData.count > 28 else { // 12 nonce + 16 tag minimum
            throw EncryptionError.invalidData
        }

        let symmetricKey = SymmetricKey(data: key)

        guard let sealedBox = try? AES.GCM.SealedBox(combined: encryptedData) else {
            throw EncryptionError.invalidData
        }

        guard let decryptedData = try? AES.GCM.open(sealedBox, using: symmetricKey) else {
            throw EncryptionError.decryptionFailed
        }

        return decryptedData
    }

    // MARK: - Sync Blob

    /// Encrypt full dataset for cloud sync (zero-knowledge)
    func encryptForSync(_ jsonData: Data, syncKey: Data) throws -> (blob: Data, checksum: String) {
        let encrypted = try encrypt(jsonData, with: syncKey)
        let checksum = sha256Hash(of: encrypted)
        return (encrypted, checksum)
    }

    /// Decrypt blob received from cloud sync
    func decryptFromSync(_ blob: Data, syncKey: Data, expectedChecksum: String) throws -> Data {
        let actualChecksum = sha256Hash(of: blob)
        guard actualChecksum == expectedChecksum else {
            throw EncryptionError.invalidData
        }
        return try decrypt(blob, with: syncKey)
    }

    // MARK: - Hashing

    /// SHA-256 hash for integrity verification
    func sha256Hash(of data: Data) -> String {
        let digest = SHA256.hash(data: data)
        return digest.map { String(format: "%02x", $0) }.joined()
    }
}

File: shifai-android/app/src/main/java/com/shifai/data/encryption/EncryptionManager.kt
package com.shifai.data.encryption

import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

/**
 * Encryption Manager — AES-256-GCM + PBKDF2 key derivation
 * Spike S0-1: IMPLEMENTED — zero-knowledge architecture
 *
 * Mirrors iOS EncryptionManager.swift for cross-platform parity
 */
class EncryptionManager {

    companion object {
        private const val PBKDF2_ITERATIONS = 100_000
        private const val KEY_LENGTH = 256 // bits
        private const val SALT_LENGTH = 32 // bytes
        private const val GCM_NONCE_LENGTH = 12 // bytes
        private const val GCM_TAG_LENGTH = 128 // bits
        private const val AES_GCM_CIPHER = "AES/GCM/NoPadding"
        private const val PBKDF2_ALGORITHM = "PBKDF2WithHmacSHA256"
    }

    sealed class EncryptionError(message: String) : Exception(message) {
        data object KeyDerivationFailed : EncryptionError("Échec de la dérivation de clé")
        data object EncryptionFailed : EncryptionError("Échec du chiffrement")
        data object DecryptionFailed : EncryptionError("Échec du déchiffrement")
        data object InvalidData : EncryptionError("Données invalides")
        data object InvalidKeyLength : EncryptionError("Longueur de clé invalide")
    }

    private val secureRandom = SecureRandom()

    // ─── Key Derivation (PBKDF2) ───

    /**
     * Derive master key from PIN using PBKDF2-SHA256
     * @param password User's PIN as char array (cleared after use)
     * @param salt Random 32-byte salt
     * @return 256-bit derived key
     */
    fun deriveMasterKey(password: CharArray, salt: ByteArray): ByteArray {
        return try {
            val spec = PBEKeySpec(password, salt, PBKDF2_ITERATIONS, KEY_LENGTH)
            val factory = SecretKeyFactory.getInstance(PBKDF2_ALGORITHM)
            factory.generateSecret(spec).encoded.also {
                spec.clearPassword()
            }
        } catch (e: Exception) {
            throw EncryptionError.KeyDerivationFailed
        }
    }

    /**
     * Derive master key from Data (for API parity with iOS)
     */
    fun deriveMasterKey(password: ByteArray, salt: ByteArray): ByteArray {
        val chars = CharArray(password.size) { password[it].toInt().toChar() }
        return deriveMasterKey(chars, salt)
    }

    // ─── Salt / Key Generation ───

    fun generateSalt(): ByteArray {
        val salt = ByteArray(SALT_LENGTH)
        secureRandom.nextBytes(salt)
        return salt
    }

    fun generateRandomKey(): ByteArray {
        val key = ByteArray(32) // 256 bits
        secureRandom.nextBytes(key)
        return key
    }

    // ─── AES-256-GCM Encrypt ───

    /**
     * Encrypt using AES-256-GCM
     * Output format: nonce (12 bytes) + ciphertext + GCM tag (16 bytes)
     * Compatible with iOS CryptoKit AES.GCM combined format
     */
    fun encrypt(plaintext: ByteArray, key: ByteArray): ByteArray {
        if (key.size != 32) throw EncryptionError.InvalidKeyLength

        return try {
            // Generate random 12-byte nonce
            val nonce = ByteArray(GCM_NONCE_LENGTH)
            secureRandom.nextBytes(nonce)

            val cipher = Cipher.getInstance(AES_GCM_CIPHER)
            val keySpec = SecretKeySpec(key, "AES")
            val gcmSpec = GCMParameterSpec(GCM_TAG_LENGTH, nonce)
            cipher.init(Cipher.ENCRYPT_MODE, keySpec, gcmSpec)

            val ciphertextAndTag = cipher.doFinal(plaintext)

            // Combine: nonce + ciphertext + tag (matches iOS format)
            nonce + ciphertextAndTag
        } catch (e: EncryptionError) {
            throw e
        } catch (e: Exception) {
            throw EncryptionError.EncryptionFailed
        }
    }

    // ─── AES-256-GCM Decrypt ───

    /**
     * Decrypt AES-256-GCM data
     * Input format: nonce (12 bytes) + ciphertext + GCM tag (16 bytes)
     */
    fun decrypt(ciphertext: ByteArray, key: ByteArray): ByteArray {
        if (key.size != 32) throw EncryptionError.InvalidKeyLength
        if (ciphertext.size <= GCM_NONCE_LENGTH + GCM_TAG_LENGTH / 8) {
            throw EncryptionError.InvalidData
        }

        return try {
            // Extract nonce (first 12 bytes)
            val nonce = ciphertext.copyOfRange(0, GCM_NONCE_LENGTH)
            val encryptedData = ciphertext.copyOfRange(GCM_NONCE_LENGTH, ciphertext.size)

            val cipher = Cipher.getInstance(AES_GCM_CIPHER)
            val keySpec = SecretKeySpec(key, "AES")
            val gcmSpec = GCMParameterSpec(GCM_TAG_LENGTH, nonce)
            cipher.init(Cipher.DECRYPT_MODE, keySpec, gcmSpec)

            cipher.doFinal(encryptedData)
        } catch (e: EncryptionError) {
            throw e
        } catch (e: Exception) {
            throw EncryptionError.DecryptionFailed
        }
    }

    // ─── Sync Operations ───

    fun encryptForSync(jsonData: ByteArray, syncKey: ByteArray): Pair<ByteArray, String> {
        val encrypted = encrypt(jsonData, syncKey)
        val checksum = sha256Hex(encrypted)
        return Pair(encrypted, checksum)
    }

    fun decryptFromSync(blob: ByteArray, syncKey: ByteArray, expectedChecksum: String): ByteArray {
        val actualChecksum = sha256Hex(blob)
        if (actualChecksum != expectedChecksum) throw EncryptionError.InvalidData
        return decrypt(blob, syncKey)
    }

    // ─── SHA-256 ───

    fun sha256Hex(data: ByteArray): String {
        val digest = MessageDigest.getInstance("SHA-256")
        return digest.digest(data).joinToString("") { "%02x".format(it) }
    }
}
